### PR TITLE
Allow folding of Stable functions when query has Sublinks

### PR DESCRIPTION
**Pull request by 24nishant**
OBJECTIVE
---------------------
Through this PR, aim is to allow folding of stable functions when the query has sublinks.
Earlier in [PR-5971](https://github.com/greenplum-db/gpdb/pull/5971), the folding of stable functions was restricted, it was only allowed if the query has RTEs. 

PROBLEM 
---------------------
It was reported that the following query is crashing in ORCA 

```
CREATE OR REPLACE FUNCTION any_func(p_dt date)
      RETURNS date
      LANGUAGE sql
      IMMUTABLE
AS $$
select (date_trunc('month', p_dt) + interval '1 month')::date;
$$
EXECUTE ON ANY;
```
```
create TABLE test_partitioned_2024 ( nagreementid int4 NULL, dtrepdate date NULL)
DISTRIBUTED BY (nagreementid)
PARTITION BY RANGE(dtrepdate)
(
    START ('2024-01-01'::date) END ('2024-12-31'::date) EVERY ('1 mon'::interval)
);

```
```
select coalesce(
                (select max(dtRepDate) from test_partitioned_2024 where dtRepDate between
                    date_trunc('month',current_date)::date  and any_func(current_date)),
                (select max(dtRepDate) from test_partitioned_2024)
               );

```

CAUSE OF FAILURE 
---------------------
1) For the above query, parsed query has no RTE but has Sublink. 
2) Prior to the fix, in such a scenario folding(to a const or inlining) of stable functions was restricted. 
3) Thus,  ** 'any_func(current_date)' in the query is not folded.**
    3.1) Not folded to constant - as restricted via check in PR-5971. any_func() has date_trunc() which is stable, and is not folded.
    3.2) Not inlined - as any_func() is an immutable function which contains a stable function date_trunc(). Thus violates condition for Inlining. 
    3.3) Thus in the algebrized query, a function call is present and it looks like - 

            |     |     +--CScalarCmp (<=)
            |     |        |--CScalarIdent "dtrepdate" (2)
            |     |        +--CScalarFunc (any_func)
            |     |           +--CScalarCoerceViaIO
            |     |              +--CScalarConst (6E6F7700)

4) When the query is optimized, **a worker is registerd in the worker pool manager (say W1 for this task)**. During dxl to planned statment translation, the translation of partition selectors is done. During this time to find the bounds for static parition elimination, the Quals in the query are evaluted. Since function call is present in the Quals, the function call is evaluated first. 

5) During this evaluation, optimization of any_func(current_date) is inititated, which fails, during query to dxl stage because of presence of T_PARAMs. This optimization is performed by the planner.

6) Note, **before initiating the above optimization (point-5), a new worker is registerd (say W2). ORCA only supports one worker, 
as a new worker is registered, the track of the old worker (W1), carrying out the initial task is lost** 
    
7) As ORCA return to the normal flow of dxl to planned statment translation after optimizing the quals, now no worker is present.
**Now whenever a worker is accessed, it is like access a nullptr, which cause the crash.** 

8) In the case of  query this crash happens to be while checking trace flags.

**POST FIX**

Now that, folding is allowed if Sublink is present, thus the function call is folded to a constant. Thus, another round of query optimization during 
static partition elimination is not performed. 

Now, ORCA is producing a plan. 

```
                                                                              QUERY PLAN
----------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Result  (cost=0.00..904755290.84 rows=1 width=4)
   ->  Nested Loop Left Join  (cost=0.00..904755290.84 rows=2 width=8)
         Join Filter: true
         ->  Nested Loop Left Join  (cost=0.00..882688.09 rows=1 width=4)
               Join Filter: true
               ->  Result  (cost=0.00..0.00 rows=1 width=1)
               ->  Materialize  (cost=0.00..431.00 rows=1 width=4)
                     ->  Aggregate  (cost=0.00..431.00 rows=1 width=4)
                           ->  Gather Motion 3:1  (slice2; segments: 3)  (cost=0.00..431.00 rows=1 width=4)
                                 ->  Sequence  (cost=0.00..431.00 rows=1 width=4)
                                       ->  Partition Selector for test_partitioned_2024 (dynamic scan id: 1)  (cost=10.00..100.00 rows=34 width=4)
                                             Partitions selected: 2 (out of 12)
                                       ->  Dynamic Seq Scan on test_partitioned_2024 test_partitioned_2024_1 (dynamic scan id: 1)  (cost=0.00..431.00 rows=1 width=4)
                                             Filter: ((dtrepdate >= '2024-02-01'::date) AND (dtrepdate <= '2024-03-01'::date))
         ->  Materialize  (cost=0.00..431.00 rows=1 width=4)
               ->  Aggregate  (cost=0.00..431.00 rows=1 width=4)
                     ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..431.00 rows=1 width=4)
                           ->  Sequence  (cost=0.00..431.00 rows=1 width=4)
                                 ->  Partition Selector for test_partitioned_2024 (dynamic scan id: 2)  (cost=10.00..100.00 rows=34 width=4)
                                       Partitions selected: 12 (out of 12)
                                 ->  Dynamic Seq Scan on test_partitioned_2024 (dynamic scan id: 2)  (cost=0.00..431.00 rows=1 width=4)
 Optimizer: Pivotal Optimizer (GPORCA)
(22 rows)
```

### CODE REVIEW

**Comment by chrishajas** [comment]
Does this also affect 7X?

File: src/backend/gporca/libgpos/src/_api.cpp [hunk 1]
```diff
@@ -173,6 +173,18 @@ gpos_exec(gpos_exec_params *params)
...
+		if (pwpm->Self())
+		{
+			// Raise an exception, if a worker already exists in the Worker Pool
+			// Manager. Since, only one worker is supported, if a new worker is
+			// registered then the address of old worker will be lost and would
+			// lead to an undefined behaviour.
+			GPOS_ASSERT(
+				false &&
+				"New worker cannot be created, as a worker already exists. ORCA supports only one worker.");
+			GPOS_RAISE(CException::ExmaUnhandled, CException::ExmiUnhandled);
```
Comment (chrishajas) [inline src/backend/gporca/libgpos/src/_api.cpp:185]: I think this should be in a separate commit. The same issue also exists in 7X, so we'll want this there too. Also, can we use a better exception with a message.

File: src/backend/optimizer/util/clauses.c [hunk 2]
```diff
@@ -5876,13 +5876,17 @@ should_eval_stable_functions(PlannerInfo *root)
 	 * Without PlannerGlobal, we cannot mark the plan as a `oneoffPlan`
 	 */
 	if (root == NULL) return false;
-	if (root->glob == NULL) return false;
+	if (root->glob == NULL || root->parse == NULL) return false;
```
Comment (khannaekta) [inline src/backend/optimizer/util/clauses.c:5879]: Why do we add this check `root->parse == NULL`?

Comment (24nishant) [reply src/backend/optimizer/util/clauses.c:5879]: Since I was accessing `root->parse->hasSubLinks`, so I first checked if `root->parse` exists.
Just  checking before accessing.

**Comment by 24nishant** [comment]
> Does this also affect 7X?

Yes, it exists in 7x also. For the same query ORCA does not crash but falls back to planner in this case. 
I tried the fix in 7x, it works and ORCA generates a plan.

File: src/backend/gporca/libgpos/src/_api.cpp (hunk 1 above)
Comment (24nishant) [reply src/backend/gporca/libgpos/src/_api.cpp]: I have updated the exception and error message.
Can you please share how can I see the error message mentioned in GPOS_RAISE()?

File: src/test/regress/sql/gporca.sql [hunk 3]
```diff
@@ -3353,6 +3353,70 @@ EXPLAIN (VERBOSE, COSTS OFF)
...
+DECLARE
+    fix_date DATE := '2024-02-10';
+BEGIN
+    RETURN fix_date;
+END;
+$$
+LANGUAGE plpgsql
+STABLE;
+
+explain select coalesce((select max(dtRepDate) from test_partitioned_2024
```
Comment (khannaekta) [inline src/test/regress/sql/gporca.sql:3386]: I think we can further simplify the query to test this scenario:
```
explain select (select dtRepDate from test_partitioned_2024                                                                                                                                                                                                                                     where dtRepDate = any_func(ret_date()));
```

**Comment by chrishajas** [comment]
> > Does this also affect 7X?
> 
> Yes, it exists in 7x also. For the same query ORCA does not crash but falls back to planner in this case. I tried the fix in 7x, it works and ORCA generates a plan.

If this issue also exists in 7X, we should create PRs for 7X first and then backport to 6X

File: src/backend/gporca/libgpos/src/_api.cpp (hunk 1 above)
Comment (chrishajas) [reply src/backend/gporca/libgpos/src/_api.cpp]: If you set `optimizer_trace_fallback=on` does that show it?

We can also use a different exception type rather than unhandled if you think that might be better (we declare some in `EresExceptionInit`)

Comment (24nishant) [reply src/backend/gporca/libgpos/src/_api.cpp]: 1. I used the flag, but In the retail build no exception message is printed. 
2. In the debug build, the message is printed. 
3. I also tried adding error message like 
		```
if (GPOS_MATCH_EX(ex, CException::ExmaInvalid, CException::ExmiORCAInvalidState2))
		{
			errstart(ERROR, ex.Filename(), ex.Line(), NULL, TEXTDOMAIN);
			errfinish(
				errcode(ERRCODE_INTERNAL_ERROR),
				errmsg(
					"Worker1 is already registered. This is an invalid state, please report this error. "));
		}
```
Only in this case, the messages if printed, but then we don't fallback to planner

File: src/test/regress/sql/gporca.sql (hunk 3 above)
Comment (24nishant) [reply src/test/regress/sql/gporca.sql:3386]: Should I add this one and keep the customer query also ?
I want to retain the original customer query.

**Comment by 24nishant** [comment]
> > > Does this also affect 7X?
> > 
> > 
> > Yes, it exists in 7x also. For the same query ORCA does not crash but falls back to planner in this case. I tried the fix in 7x, it works and ORCA generates a plan.
> 
> If this issue also exists in 7X, we should create PRs for 7X first and then backport to 6X

Yes, true but I did not raise the 7x PR first because when we received the failing query, it was crashing in 6X but falling back in 7X.  Apparently the cause also seemed to be different, so all effort was focussed towards solving 6x first.  
It's after solving the 6x case, when i checked the same fix on 7x , i found that ORCA now generates a plan here also.
I will raise the 7x PR next.

File: src/test/regress/sql/gporca.sql (hunk 3 above)
Comment (khannaekta) [reply src/test/regress/sql/gporca.sql:3386]: I see. Usually we prefer to have a minimal query to allow just testing codepath we fixed and avoid any additional noise.

File: src/backend/optimizer/util/clauses.c (hunk 2 above)
Comment (khannaekta) [reply src/backend/optimizer/util/clauses.c:5879]: Interestingly we always accessed `root->parse->rtable` but never checked for `root->parse` being NULL. Good call.

File: src/test/regress/sql/gporca.sql (hunk 3 above)
Comment (chrishajas) [reply src/test/regress/sql/gporca.sql:3386]: I agree with @khannaekta. We want a minimal query that showcases this behavior. We specifically don't want to include complex user queries that would only add noise.

File: src/backend/gporca/libgpos/src/_api.cpp (hunk 1 above)
Comment (chrishajas) [reply src/backend/gporca/libgpos/src/_api.cpp]: Ah, ok. I think even if the exception isn't printed, falling back is still better than crashing so I'd rather keep the defensive check in (though it'd be ideal to also print a fallback message).

**Comment by THANATOSLAVA** [comment]
A few thoughts:

1. I am not quite convinced that we took the correct approach here. Function folding is an optimization technique. We shouldn't crash because we fail to optimize. This is especially the case with stable functions. From what I gather, Postgres doesn't pre-evaluate stable functions. GPDB does do so as a hack for better performance. We shouldn't need to relying on a workaround to avoid crashing.
2. I think RCA bullets 4~7 need to be rephrased. I'm guessing the scenario that you tried to describe is --
The main query (planned by ORCA) is dispatched onto segments. Now the segment sees the function and attempts to evaluate but fails. So instead of returning a const to the main query, the segment attempts to dispatch the function. And that caused the error?
This leads to the second reason why I am skeptical that `should_eval_stable_functions` is where the changes should happen. Both 6X and 7X planner managed to evaluate `any_func` before your changes. Wouldn't that indicate the issue is with the ORCA rather than the planner code?
3. I don't think this issue has to do with function inlining. Looks like the goal is to get `any_func` to return a value, rather than being planned as part of the main query.

**Comment by 24nishant** [comment]
> A few thoughts:
> 
> 1. I am not quite convinced that we took the correct approach here. Function folding is an optimization technique. We shouldn't crash because we fail to optimize. This is especially the case with stable functions. From what I gather, Postgres doesn't pre-evaluate stable functions. GPDB does do so as a hack for better performance. We shouldn't need to relying on a workaround to avoid crashing.
> 2. I think RCA bullets 4~7 need to be rephrased. I'm guessing the scenario that you tried to describe is --
>    The main query (planned by ORCA) is dispatched onto segments. Now the segment sees the function and attempts to evaluate but fails. So instead of returning a const to the main query, the segment attempts to dispatch the function. And that caused the error?
>    This leads to the second reason why I am skeptical that `should_eval_stable_functions` is where the changes should happen. Both 6X and 7X planner managed to evaluate `any_func` before your changes. Wouldn't that indicate the issue is with the ORCA rather than the planner code?
> 3. I don't think this issue has to do with function inlining. Looks like the goal is to get `any_func` to return a value, rather than being planned as part of the main query.

Thanks @THANATOSLAVA, for sharing these thoughts. 

1. I checked the execution of Planner for this query, found that in the preprocessing, Planner too evaluates the function call to compute a constant value. Please check the `eval_const_expressions()` [here](https://github.com/greenplum-db/gpdb/blob/6f30fead0851411a0f5639cb1f9277d37cc746f1/src/backend/optimizer/util/clauses.c#L2539) .
2. From `eval_const_expressions()`, `eval_const_expressions_mutator()` is called which gives the processed node back
3. Similarly, if we call [`fold_constants()`](https://github.com/greenplum-db/gpdb/blob/6f30fead0851411a0f5639cb1f9277d37cc746f1/src/backend/optimizer/util/clauses.c#L2407C1-L2407C15), it also calls `eval_const_expressions_mutator`, please check [here](https://github.com/greenplum-db/gpdb/blob/6f30fead0851411a0f5639cb1f9277d37cc746f1/src/backend/optimizer/util/clauses.c#L2426C7-L2426C37)
4. Thus,  both ORCA and planner are evaluating the functions.

File: src/backend/gporca/libgpos/src/_api.cpp (hunk 1 above)
Comment (24nishant) [reply src/backend/gporca/libgpos/src/_api.cpp]: I have corrected it, now we have a fallback message in retail build also

File: src/test/regress/sql/gporca.sql (hunk 3 above)
Comment (24nishant) [reply src/test/regress/sql/gporca.sql]: Thank you @khannaekta & @chrishajas for guiding me, I have updated the query.

**Comment by 24nishant** [comment]
Closing the PR, to find how planner optimizes the query mentioned in the description.
Will then discuss & decide on whether we need an alternate approach to solve.
Thank you @chrishajas , @khannaekta  and @THANATOSLAVA  for the feedbacks.
